File: frontend/src/api/loaImports.ts
import { apiRequest } from "./client";

export type LoaImportLine = {
  id: string; line_number: number; product_id?: string;
  candidate_key?: string; source_order?: number; group_id?: string;
  description?: string; description_raw?: string; description_normalized?: string;
  hsn_text?: string; hsn_code_id?: string; unit_text?: string; uom_raw?: string; uom_normalized?: string;
  unit_id?: string; quantity?: string; rate?: string; amount?: string;
  oem_make?: string; model_number?: string; tax_text?: string; remarks?: string;
  source_page?: number; source_page_start?: number; source_page_end?: number;
  source_serial?: string; source_raw_text?: string; extraction_method?: string;
  extraction_confidence?: string; extraction_issues?: string[];
  extraction_outcome: "EXTRACTED" | "NEEDS_REVIEW" | "REJECTED_WITH_REASON" | "EXPLICITLY_IGNORED_BY_OWNER";
  extraction_issue?: string;
};
export type LoaImportGroup = { id:string; sequence:number; source_key:string; title_raw:string; title_normalized:string; source_kind:string; source_page_start?:number; source_page_end?:number; source_total?:string; extracted_total?:string; difference?:string; reconciliation_status:string };
export type LoaImportSchedule = { id:string; sequence:number; source_key:string; title_raw:string; title_normalized:string; source_page_start?:number; source_page_end?:number; source_total?:string; extracted_total?:string; difference?:string; reconciliation_status:string; groups:LoaImportGroup[] };
export type LoaImport = {
  id: string; original_filename: string; mime_type: string; extension: string;
  size_bytes: number; uploaded_by_user_id: string; uploaded_at: string; status: string;
  extraction_method?: string; extraction_error?: string; extraction_warnings: string[];
  project_id?: string; railway_zone_id?: string; railway_division_id?: string;
  authority_id?: string; issuing_party_id?: string; loa_id?: string;
  extracted_zone_text?: string; extracted_division_text?: string; authority_text?: string;
  loa_number?: string; tender_reference?: string; loa_date?: string;
  completion_period?: string; completion_date?: string; work_description?: string;
  contract_value?: string; duplicate_candidates: Array<{id:string;loa_number:string;project_id:string}>;
  project_candidates: Array<{id:string;code:string;name:string;reason:string}>;
  authority_candidates: Array<{text:string;role:string;source:string;master_id?:string;master_status?:"MATCHED"|"NOT_CONFIGURED"|"AMBIGUOUS";master_detail?:string}>;
  boq_reconciliation: Record<string, unknown>;
  boq_readiness_issues: Array<{scope:string;line_number?:number;field:string;message:string}>;
  completion_date_provenance?: "SOURCE_EXTRACTED" | "DERIVED" | "OWNER_CORRECTED" | "WAITING_FOR_LOA_DATE";
  loa_date_provenance?: "SOURCE_EXTRACTED" | "OWNER_CORRECTED";
  loa_date_source?: string;
  lines: LoaImportLine[];
  schedules?: LoaImportSchedule[];
};

export function uploadRailwayLoa(token: string, file: File): Promise<LoaImport> {
  const body = new FormData(); body.append("file", file);
  return apiRequest<LoaImport>("/railway-loa-imports", { method: "POST", token, body });
}
export const listLoaImports = (token: string) => apiRequest<LoaImport[]>("/railway-loa-imports", { token });
export const getLoaImport = (token: string, id: string) => apiRequest<LoaImport>(`/railway-loa-imports/${id}`, { token });
export const saveLoaImportReview = (token: string, id: string, data: Record<string, unknown>) => apiRequest<LoaImport>(`/railway-loa-imports/${id}`, { method: "PATCH", token, body: JSON.stringify(data) });
export const retryLoaImport = (token: string, id: string) => apiRequest<LoaImport>(`/railway-loa-imports/${id}/retry`, { method: "POST", token });
export const resolveLoaImportMasters = (token: string, id: string) => apiRequest<LoaImport>(`/railway-loa-imports/${id}/resolve-masters`, { method: "POST", token });
export const mapRailwayCustomer = (token: string, id: string, customerPartyId: string) => apiRequest<LoaImport>(`/railway-loa-imports/${id}/customer-mapping`, { method: "POST", token, body: JSON.stringify({ customer_party_id: customerPartyId }) });
export const approveLoaImport = (token: string, id: string, data: Record<string, unknown> = {}) => apiRequest<LoaImport>(`/railway-loa-imports/${id}/approve`, { method: "POST", token, body: JSON.stringify(data) });
export const cancelLoaImport = (token: string, id: string) => apiRequest<LoaImport>(`/railway-loa-imports/${id}/cancel`, { method: "POST", token });

export async function openOriginalLoa(token: string, id: string, filename: string): Promise<void> {
  const response = await fetch(`${import.meta.env.VITE_API_BASE_URL ?? "/api/v1"}/railway-loa-imports/${id}/original`, { headers: { Authorization: `Bearer ${token}` } });
  if (!response.ok) throw new Error("Unable to open original LOA.");
  const url = URL.createObjectURL(await response.blob());
  const anchor = document.createElement("a"); anchor.href = url; anchor.target = "_blank"; anchor.download = filename.endsWith(".xlsx") ? filename : ""; anchor.click();
  setTimeout(() => URL.revokeObjectURL(url), 30_000);
}

File: backend/app/schemas/loa_imports.py
from datetime import date, datetime
from decimal import Decimal
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field

from app.schemas.contracts import ProjectCreate


class LoaImportLineReview(BaseModel):
    id: UUID | None = None
    candidate_key: str | None = Field(default=None, max_length=100)
    source_order: int | None = Field(default=None, ge=1)
    group_id: UUID | None = None
    product_id: UUID | None = None
    description: str | None = Field(default=None, max_length=10000)
    description_raw: str | None = Field(default=None, max_length=50000)
    description_normalized: str | None = Field(default=None, max_length=50000)
    hsn_text: str | None = Field(default=None, max_length=50)
    hsn_code_id: UUID | None = None
    unit_text: str | None = Field(default=None, max_length=100)
    uom_raw: str | None = Field(default=None, max_length=100)
    uom_normalized: str | None = Field(default=None, max_length=100)
    unit_id: UUID | None = None
    quantity: Decimal | None = Field(default=None, ge=0, max_digits=18, decimal_places=4)
    rate: Decimal | None = Field(default=None, ge=0, max_digits=18, decimal_places=2)
    amount: Decimal | None = Field(default=None, ge=0, max_digits=18, decimal_places=2)
    oem_make: str | None = Field(default=None, max_length=255)
    model_number: str | None = Field(default=None, max_length=255)
    tax_text: str | None = Field(default=None, max_length=255)
    remarks: str | None = Field(default=None, max_length=1000)
    source_page: int | None = Field(default=None, ge=1)
    source_page_start: int | None = Field(default=None, ge=1)
    source_page_end: int | None = Field(default=None, ge=1)
    source_serial: str | None = Field(default=None, max_length=50)
    source_raw_text: str | None = Field(default=None, max_length=10000)
    extraction_outcome: str = Field(
        default="EXTRACTED",
        pattern="^(EXTRACTED|NEEDS_REVIEW|REJECTED_WITH_REASON|EXPLICITLY_IGNORED_BY_OWNER)$",
    )
    extraction_issue: str | None = Field(default=None, max_length=1000)
    extraction_method: str | None = Field(default=None, max_length=20)
    extraction_confidence: Decimal | None = Field(default=None, ge=0, le=1)
    extraction_issues: list[str] = Field(default_factory=list)


class LoaImportReview(BaseModel):
    project_id: UUID | None = None
    railway_zone_id: UUID | None = None
    railway_division_id: UUID | None = None
    authority_id: UUID | None = None
    issuing_party_id: UUID | None = None
    loa_number: str | None = Field(default=None, max_length=100)
    tender_reference: str | None = Field(default=None, max_length=255)
    loa_date: date | None = None
    completion_period: str | None = Field(default=None, max_length=255)
    completion_date: date | None = None
    work_description: str | None = Field(default=None, max_length=10000)
    contract_value: Decimal | None = Field(default=None, ge=0, max_digits=18, decimal_places=2)
    authority_text: str | None = Field(default=None, max_length=500)
    lines: list[LoaImportLineReview] | None = None


class LoaImportApproval(BaseModel):
    new_project: ProjectCreate | None = None
    confirm_duplicate: bool = False


class RailwayCustomerMapping(BaseModel):
    customer_party_id: UUID


class LoaImportLineResponse(LoaImportLineReview):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    line_number: int


class LoaImportGroupResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    sequence: int
    source_key: str
    title_raw: str
    title_normalized: str
    source_kind: str
    source_page_start: int | None
    source_page_end: int | None
    source_total: Decimal | None
    extracted_total: Decimal | None
    difference: Decimal | None
    reconciliation_status: str


class LoaImportScheduleResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    sequence: int
    source_key: str
    title_raw: str
    title_normalized: str
    source_page_start: int | None
    source_page_end: int | None
    source_total: Decimal | None
    extracted_total: Decimal | None
    difference: Decimal | None
    reconciliation_status: str
    groups: list[LoaImportGroupResponse]


class LoaImportResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    original_filename: str
    mime_type: str
    extension: str
    size_bytes: int
    sha256: str
    uploaded_by_user_id: UUID
    status: str
    extraction_method: str | None
    extraction_error: str | None
    extraction_warnings: list
    project_id: UUID | None
    railway_zone_id: UUID | None
    railway_division_id: UUID | None
    authority_id: UUID | None
    issuing_party_id: UUID | None
    loa_id: UUID | None
    extracted_zone_text: str | None
    extracted_division_text: str | None
    authority_text: str | None
    loa_number: str | None
    tender_reference: str | None
    loa_date: date | None
    completion_period: str | None
    completion_date: date | None
    work_description: str | None
    contract_value: Decimal | None
    duplicate_candidates: list
    project_candidates: list
    authority_candidates: list
    boq_reconciliation: dict
    boq_readiness_issues: list[dict] = Field(default_factory=list)
    completion_date_provenance: str | None
    loa_date_provenance: str | None
    loa_date_source: str | None
    uploaded_at: datetime
    extracted_at: datetime | None
    approved_at: datetime | None
    updated_at: datetime
    lines: list[LoaImportLineResponse]
    schedules: list[LoaImportScheduleResponse] = Field(default_factory=list)
